pkg/ignore/types.go

```diff
@@ -0,0 +1,162 @@
+
+	// convenience to flag the current path as not root directory
+	isRootDir := path == directory
+
+	// dont append the root directory as its always included
+	if isRootDir {
+		return No, nil
+	}
+
+	// if the dir is any of default patterns then skip scanning the dir
+	skipPaths := sets.NewString(defaultPatterns...)
+	if skipPaths.Has(fi.Name()) {
+		return Transitive, nil
+	}
+
+	// flag to keep track if file or directory is excluded
+	isExcluded := false
+
+	// flag to keep track transitive  directories
+	// Transitive directories are are the ones that are excluded
+	// but might have files under them with inversions
+	// e.g. with a ignore file like
+	// foo
+	// !foo/bar/one.txt
+	// In above case directory foo is transitive
+	hasTransitives := false
+
+	for _, igp := range ignorePatterns {
+
+		re, err := regexp.Compile(igp.regExpr)
+		// if the regular expression fails to compile then skip to next pattern
+		if err != nil {
+			isExcluded = false
+			continue
+		}
+
+		// check if the parent path is an excluded pattern in the list
+		parentDirMatch := excludedDirs.Has(filepath.Dir(path))
+		// if the parent directory is not in the list check if it matches
+		// any pattern
+		if !parentDirMatch {
+			parentDirMatch = re.MatchString(filepath.Dir(path))
+		}
+		// if pattern is not a inversion and the excluded directories has
+		// the current path parent the continue to check next pattern
+		if !igp.invert && parentDirMatch {
+			isExcluded = true
+			ignorable = Current
+			excludedDirs.Insert(filepath.Dir(path))
+			continue
+		}
+
+		regxMatches := re.FindAllStringSubmatch(path, -1)
+
+		if len(regxMatches) > 0 {
+			isExcluded = true
+			// when a file or directory matches the pattern but has inversion
+			// then add the file to include list
+			if igp.invert {
+				ignorable = No
+			} else {
+				if fi.IsDir() {
+					excludedDirs.Insert(path)
+				}
+				ignorable = Current
+			}
+		}
+
+		if fi.IsDir() && !isRootDir && igp.invert && igp.paths != nil {
+			pPath := strings.Join(igp.paths, string(os.PathSeparator))
+			pPath = filepath.Join(directory, pPath)
+			if path == pPath {
+				hasTransitives = true
+			}
+		}
+
+	}
+	// the directory is not transitive and not root directory check is being done
+	// at last to avoid skipping directories that are not listed in the ignore file
+	// with "!" i.e. implicit includes
+	if fi.IsDir() && !isRootDir && isExcluded && !hasTransitives {
+		ignorable = Transitive
+	}
+
+	return ignorable, nil
+}
+
+// defaultIgnorer is the default FileIgnorer which is returned when no .dockerignore file is present
+// or error processing .dockerignore
+type defaultIgnorer struct {
+	// root directory to start the scanner
+	directory    string
+	excludedDirs sets.String
+}
+
+// CanIgnore is the directory scanner to scan directory for standard ignore patterns
+func (i *defaultIgnorer) CanIgnore(path string, fi os.FileInfo) (Ignorable, error) {
+	return scanDir(i.directory, path, fi, i.excludedDirs)
```

Comment:
Feels like the bulk of `scanDir` doesn't benefit the default ignorer, and it is keying off of the global `ignorePatterns` state, which should be a property of the dockerIgnorer.  Maybe inline/duplicate the first handful of lines and make the rest of `scanDir` a member of dockerIgnorer

Reply:
Actually I was contemplating on this change and was working on my local copy thought to add it with the next review updates :), now adding it.